pymatgen/electronic_structure/dos.py

```diff
@@ -1165,6 +1166,166 @@ def get_upper_band_edge(
+            (default is summed_pdos)
+            min_e (float): The minimum mode energy to include in the fingerprint (default is None)
+            max_e (float): The maximum mode energy to include in the fingerprint (default is None)
+            n_bins (int): Number of bins to be used in the fingerprint (default is 256)
+            normalize (bool): If true, normalizes the area under fp to equal to 1 (default is True)
+
+        Returns:
+            Fingerprint(namedtuple) : The electronic density of states fingerprint
+            of format (energies, densities, type, n_bins)
+        """
+        fingerprint = namedtuple("fingerprint", "energies densities type n_bins bin_width")
+        energies = self.energies - self.efermi
+
+        if max_e is None:
+            max_e = np.max(energies)
+
+        if min_e is None:
+            min_e = np.min(energies)
+
+        pdos_obj = self.get_spd_dos()
+
+        pdos = {}
+        for k, _v in pdos_obj.items():
+            dens = pdos_obj[k].get_densities()
+
+            pdos.update({k.name: dens})
+
+        sum_pdos_array = []
+        for dos in pdos.values():
+            sum_pdos_array.append(np.array(dos))
+
+        pdos.update({"summed_pdos": np.sum(sum_pdos_array, axis=0)})
+        pdos.update({"tdos": self.get_densities()})
+
+        try:
+
+            densities = pdos[type]
+            if len(energies) < n_bins:
+                inds = np.where((energies >= min_e) & (energies <= max_e))
+                return fingerprint(energies[inds], densities[inds], type, len(energies), np.diff(energies)[0])
+
+            if binning:
+                ener_bounds = np.linspace(min_e, max_e, n_bins + 1)
+                ener = ener_bounds[:-1] + (ener_bounds[1] - ener_bounds[0]) / 2.0
+                bin_width = np.diff(ener)[0]
+            else:
+                ener_bounds = np.array(energies)
+                ener = np.append(energies, [energies[-1] + np.abs(energies[-1]) / 10])
+                n_bins = len(energies)
+                bin_width = np.diff(energies)[0]
+
+            dos_rebin = np.zeros(ener.shape)
+
+            for ii, e1, e2 in zip(range(len(ener)), ener_bounds[0:-1], ener_bounds[1:]):
+                inds = np.where((energies >= e1) & (energies < e2))
+                dos_rebin[ii] = np.sum(densities[inds])
+            if normalize:  # scale dos bins to make area under histogram equal 1
+                area = np.sum(dos_rebin * bin_width)
+                dos_rebin_sc = dos_rebin / area
+            else:
+                dos_rebin_sc = dos_rebin
+
+            return fingerprint(np.array([ener]), dos_rebin_sc, type, n_bins, bin_width)
+
+        except KeyError:
+            raise ValueError(
+                "Please recheck type requested, either the orbital projections unavailable in input dos or "
+                "some there exist some mistake in the spelling."
+            )
+
+    @staticmethod
+    def fp_to_dict(fp):
```

Comment:
Missing type annotation

Reply:
I have added the type annotation (namedtuple) here now.